File: C++/decorator/Tester.cpp
#include "iostream"
#include "vector"

#include "./pizza/FarmHouse.h"
#include "./pizza/Margherita.h"
#include "./pizza/VegDelight.h"

#include "./toppings/Jalapeno.h"
#include "./toppings/Mushroom.h"
#include "./toppings/ExtraCheese.h"


using namespace std;

int main() {
    vector<BasePizza*> pizzas;

    pizzas.push_back(new Margherita()); // Margherita Pizza
    pizzas.push_back(new ExtraCheese(new Margherita())); // Extra Cheese Margherita Pizza
    pizzas.push_back(new Jalpeno(new Margherita())); // Jalapenos Margherita Pizza
    pizzas.push_back(new ExtraCheese(new Jalpeno(new Margherita()))); // Extra Cheese Jalapenos Margherita Pizza
    pizzas.push_back(new ExtraCheese(new ExtraCheese(new Jalpeno(new Margherita())))); // Dual Extra Cheese Jalapenos Margherita Pizza

    // Printing the costs
    for(int i = 0; i < pizzas.size(); i++) {
        BasePizza* basePizza = pizzas[i];
        cout << "Pizza " << i << ":\n";
        cout << basePizza->cost() << endl << endl;
    }
    
    return 0;
}

/*
Output:

Pizza 0:
Adding the cost for Margherita!
150

Pizza 1:
Adding the cost for ExtraCheese!
Adding the cost for Margherita!
170

Pizza 2:
Adding the cost for Jalapenos!
Adding the cost for Margherita!
165

Pizza 3:
Adding the cost for ExtraCheese!
Adding the cost for Jalapenos!
Adding the cost for Margherita!
185

Pizza 4:
Adding the cost for ExtraCheese!
Adding the cost for ExtraCheese!
Adding the cost for Jalapenos!
Adding the cost for Margherita!
205

*/

File: Java/decoratorpattern/Tester.java
package decoratorpattern;

import decoratorpattern.pizza.Cheese;
import decoratorpattern.pizza.BasePizza;
import decoratorpattern.pizza.FarmHouse;
import decoratorpattern.pizza.VegDelight;
import decoratorpattern.toppingdecorator.ExtraCheese;
import decoratorpattern.toppingdecorator.Jalepeno;
import decoratorpattern.toppingdecorator.Mushroom;

import java.util.ArrayList;
import java.util.List;

public class Tester {

    public static void main(String[] args) {

        List<BasePizza> pizzas = new ArrayList<>();

        pizzas.add(new Mushroom(new Jalepeno(new Cheese())));
        pizzas.add(new Mushroom(new ExtraCheese(new VegDelight())));
        pizzas.add(new Mushroom(new ExtraCheese(new Jalepeno(new ExtraCheese(new FarmHouse())))));

        for(BasePizza basePizza : pizzas) {
            System.out.println(basePizza.cost());
            System.out.println();
        }
    }

}

/*
* Output:
*
Adding the cost of Mushroom in Jalepeno!
Adding the cost of Jalepeno in Cheese!
Adding the cost of Cheeze Pizza!
150

Adding the cost of Mushroom in ExtraCheese!
Adding the cost of Extra Cheeze in VegDelight!
Adding the cost of VegDelight Pizza!
340

Adding the cost of Mushroom in ExtraCheese!
Adding the cost of Extra Cheeze in Jalepeno!
Adding the cost of Jalepeno in ExtraCheese!
Adding the cost of Extra Cheeze in FarmHouse!
Adding the cost of FarmHouse Pizza!
270
*
*
* */
